docs(contributing): Add Documentation Style Guide (#1733)
spinnaker/spinnaker.github.io: community/contributing/docs/docs-style-guide.md

---
layout: single
title:  "Documentation Style Guide"
sidebar:
  nav: community
---

{% include toc %}

This page gives writing style guidelines for the Spinnaker documentation. Since these are guidelines, not rules, use your best judgment when creating content. Feel free to propose changes to this document in a pull request.

> If you have a style question that isn't answered in this guide, ask the SIG Docs team in the `#sig-documentation` channel.

Reviewer: This assumes the reader/contributor is already familiar with the Spinnaker Slack workspace. Is that a safe assumption? (It may well be; I'm just wondering.)

For example, use
~~~~~~~~~~
See the [Repository structure](https://github.com/pf4j/pf4j-update#repository-structure) section of the PF4J README for details.
~~~~~~~~~~

rather than
~~~~~~~~~~
Click [here](https://github.com/pf4j/pf4j-update#repository-structure) to read more.
~~~~~~~~~~

For long URLs, consider using [reference-style hyperlinks](https://www.markdownguide.org/basic-syntax/#reference-style-links) to maintain readability of the Markdown file.

## Patterns to avoid

### Avoid using "we"

Do not use "we" because readers may not know if they are part of the "we".

|**Do** |**Do Not**|
|:------|:---------|
| Version 1.19.0 includes ... | In version 1.19.0, we have added ...
| Spinnaker provides a new feature for ... | We provide a new feature for ...|
| This guide teaches you how to use Plugins. | In this guide, we are going to learn about Plugins.|

### Avoid jargon and idioms

Avoid jargon and idioms to help non-native English speakers understand the content better.

|**Do** |**Do Not**|
|:------|:---------|
| Internally, ...| Under the hood, ...|
| Create a new instance. | Spin up a new instance.|

### Avoid statements about the future

> If you need to write about an alpha feature, put the text in a note or under a heading that identifies it as alpha information.

Reviewer: ```suggestion
If you need to write about an alpha feature, put the text in a note or under a heading that identifies it as alpha information, but don't include statements about when the feature will no longer be alpha, or what will change.
```

### Avoid statements that will soon be out of date

Avoid using words like "currently" and "new." A feature that is new today might not be considered new in a few months.

|**Do** |**Do Not**|
|:------|:---------|
| In version 1.4, ... | In the current version, ...|
| The Plugins feature provides ... | The new Plugin feature provides ...|

## Documentation formatting standards

### Use sentence capitalization for headings

~~~~~~~~~

> # Creating a custom webhook stage

Reviewer: I can't find it atm in the Google dev doc style guide, so maybe I'm off base, but...
I think we want to encourage imperative verbs in headings rather than gerunds, though we prefer gerunds in article titles.

### Use angle brackets for placeholders

Use angle brackets for placeholders. Describe what a placeholder represents.

For example:
`hal plugins repository add <unique-repo-name> --url <repo-url>`

### Use bold for user interface elements

|**Do** |**Do Not**|
|:------|:---------|
| Click **Fork**.| Click "Fork".|
| Select **Other**. | Select "Other".|

### Use italics to define or introduce new terms

|**Do** |**Do Not**|
|:------|:---------|
|A _Stage_ is a step in a pipeline ... | A "Stage" is a step in a pipeline ...|

### Use code style for filenames, directories, and paths

|**Do** |**Do Not**|
|:------|:---------|
| Open the `rosco.yaml` file. | Open the rosco.yaml file.|
| Go to the `/docs/tutorials` directory. | Go to the /docs/tutorials directory.|
| Open the `/.hal/config` file. | Open the /.hal/config file.|

### Lists
Group items in a list that are related to each other. Use a numbered list for instructions that need to be completed in a specific order.

 - End each item in a list with a period if one or more items in the list are complete sentences. For the sake of consistency, normally either all items or none should be complete sentences.

>  - Use the number one (1.) for ordered lists.

Reviewer: ```suggestion
 - Use the number `1.` for each item in ordered lists.
```

 - Use (+), (* ), or (-) for unordered lists.
 - Leave a blank line after each list.

>  - Indent nested lists with four spaces (for example, ⋅⋅⋅⋅).

Reviewer: Is that the correct md rule? I always just assumed it's the number of spaces required to line up exactly under the text above. in fact, for unordered lists I use three spaces (cause the * only takes up one space).

Author: I double checked online markdown resources - most state one tab or four spaces.  Brian has a point in that it depends on the Markdown converter.  The [Kramdown](https://kramdown.gettalong.org/quickref.html) docs don't state tabs or spaces.

>  - List items may consist of multiple paragraphs. Each subsequent paragraph in a list item must be indented by either four spaces or one tab.

Reviewer: Uh oh. Spaces v. tabs. 🚀 

But my same question applies. For list paragraphs, I indent with spaces to line up.

Author: So do I. For people new to markdown, though, having a concrete guideline is perhaps better.

Use meaningful variable names that have a context rather than  'foo','bar', and similar meaningless variable names.

Use a single backtick (\`) to surround inline code in a Markdown document. In Markdown:
~~~~~~~~~
Run `hal deploy apply` to deploy Spinnaker.
~~~~~~~~~

Use triple backticks to enclose a code block. In Markdown:
~~~~~~~~~
```json
{
  "firstName": "John",
  "lastName": "Smith",
  "age": 25
}
```
~~~~~~~~~

Remove trailing spaces in all code blocks.

### Don't include the command prompt in code snippets

|**Do** |**Do Not**|
|:------|:---------|
hal deploy apply| $ hal deploy apply

### Separate commands from output

Verify that the Pod is running on your chosen node:

    kubectl get pods --output=wide

The output is similar to this:

    NAME     READY     STATUS    RESTARTS   AGE    IP           NODE
    nginx    1/1       Running   0          13s    10.200.0.4   worker0

### Use normal style for string and integer field values

For field values of type string or integer, use normal style without quotation marks.

|**Do** |**Do Not**|
|:------|:---------|
Set the value of `enabled` to True. | Set the value of `enabled` to "True".

> Set the value of `image` to nginx:1.8. | Set the value of `image` to `nginx:1.8`.

Reviewer: But code-font/backtics, right?

Set the value of `enabled` to `True`. 
Set the value of `image` to `nginx:1.8`.

Author: Excellent catch - updated

## Versioning Spinnaker examples

Code examples and configuration examples that include version information should be consistent with the accompanying text.

If the information is version specific, the Spinnaker version needs to be defined in the `Prerequisites` section of the guide.

## Spinnaker.io word list

A list of Spinnaker-specific terms and words to be used consistently across the site:

| **Do**                         | **Do Not**                                                         |
|:------------------------------ |:------------------------------------------------------------------ |

> | Kubernetes                     | kubernetes                           |

Reviewer: ```suggestion
| Kubernetes                     | kubernetes, k8s                       |
```

| Docker                         | docker                               |
| SIG Docs                       | SIG Docs rather than SIG-DOCS or other variations.                 |

> | On-premises                    | On-premises or On-prem rather than On-premise or other variations. |

Reviewer: ```suggestion
| On-premises, On-prem           |  On-premise or other variations. |
```

---
layout: single
title:  "Documentation Style Guide"
sidebar:
  nav: community
---

{% include toc %}

This page gives writing style guidelines for the Spinnaker documentation. Since these are guidelines, not rules, use your best judgment when creating content. Feel free to propose changes to this document in a pull request.

If you have a style question that isn't answered in this guide, ask the SIG Docs team in the `#sig-documentation` channel.

## Language

> The Spinnaker documentation uses U.S. English spelling and grammar. However, use the international standard for punctuation inside quotes.

Reviewer 2: I don't think people would know what `the international standard for punctuation inside quotes` is. I don't.

Reviewer: I don't either, but it's illustrated below. Maybe if you just end the sentence above with a colon, you'd imply that _this thing here_ is the standard. 

Or add a link to the standard?

| Spinnaker provides a new feature for ... | We provide a new feature for ...|
| This guide teaches you how to use Plugins. | In this guide, we are going to learn about Plugins.|

### Avoid jargon and idioms

Avoid jargon and idioms to help non-native English speakers understand the content better.

|**Do** |**Do Not**|
|:------|:---------|
| Internally, ...| Under the hood, ...|
| Create a new instance. | Spin up a new instance.|

### Avoid statements about the future

If you need to write about an alpha feature, put the text in a note or under a heading that identifies it as alpha information.

### Avoid statements that will soon be out of date

Avoid using words like "currently" and "new." A feature that is new today might not be considered new in a few months.

|**Do** |**Do Not**|
|:------|:---------|
| In version 1.4, ... | In the current version, ...|
| The Plugins feature provides ... | The new Plugin feature provides ...|

## Documentation formatting standards

### Use sentence capitalization for headings

~~~~~~~~~
# Creating a custom webhook stage

## Configuring parameters for custom webhook stages
~~~~~~~~~

### Line breaks

> U#se a single newline to separate block-level content like headings, lists, images, code blocks, paragraphs, and others.

Reviewer 2: Random # in `Use`

|**Do** |**Do Not**|
|:------|:---------|
| The PipelineController restarts a Stage. | The pipeline controller restarts a stage.|
| The AmazonInfrastructureController is responsible for ... | The AmazonInfrastructureController object is responsible for ...|

### Use angle brackets for placeholders

Use angle brackets for placeholders. Describe what a placeholder represents.

For example:
`hal plugins repository add <unique-repo-name> --url <repo-url>`

### Use bold for user interface elements

|**Do** |**Do Not**|
|:------|:---------|
| Click **Fork**.| Click "Fork".|
| Select **Other**. | Select "Other".|

### Use italics to define or introduce new terms

|**Do** |**Do Not**|
|:------|:---------|
|A _Stage_ is a step in a pipeline ... | A "Stage" is a step in a pipeline ...|

### Use code style for filenames, directories, and paths

|**Do** |**Do Not**|
|:------|:---------|
| Open the `rosco.yaml` file. | Open the rosco.yaml file.|
| Go to the `/docs/tutorials` directory. | Go to the /docs/tutorials directory.|
| Open the `/.hal/config` file. | Open the /.hal/config file.|

### Lists

> Group items in a list that are related to each other. Use a numbered list for items that need to appear in a specific order.

Author: Better would be: Use a numbered list for instructions that need to be completed in a specific order.
